controllers/bucket_controller.go

```diff
@@ -74,8 +86,8 @@ type BucketProvider interface {
 	BucketExists(context.Context, string) (bool, error)
 	ObjectExists(context.Context, string, string) (bool, error)
 	FGetObject(context.Context, string, string, string) error
-	ListObjects(context.Context, gitignore.Matcher, string, string) error
 	ObjectIsNotFound(error) bool
+	VisitObjects(context.Context, string, func(string) error) error
 	Close(context.Context)
 }
```

Comment:
> Think it would be better if [the interface] moved to pkg/bucket (or internal/bucket) [...]
> This creates better separation between the reconciler and the bucket clients

I don't see how it does that. Then the controller code would be coupled to the package with the interface _and_ the packages with the implementations, since it's the controller that decides which implementation it wants (because it has access to the bucket object). In general it's better to declare requirements in the place they are required.

If all the funcs using a bucket client could be put in a package (e.g., `internal/bucket`) without creating a cycle, then it might be reasonable to move the interface (or better: more precise interfaces) there, since that's where the requirement would be. But it would be a bit of an empty gesture, when there are no other consumers of those funcs.